3rdParty/fuerte/src/HttpConnection.cpp

```diff
@@ -350,28 +395,39 @@ template <SocketType ST>
 void HttpConnection<ST>::asyncWriteNextRequest() {
   FUERTE_LOG_HTTPTRACE << "asyncWriteNextRequest: this=" << this << "\n";
   assert(_active.load());
+  if (_active == false) {
+    std::abort();
+  }
+  if (_item != nullptr) {
+    std::abort();
+  }
 
   http::RequestItem* ptr = nullptr;
   if (!_queue.pop(ptr)) {
     _active.store(false);
-    if (!_queue.pop(ptr)) {
+    if (_queue.empty()) {
       FUERTE_LOG_HTTPTRACE << "asyncWriteNextRequest: stopped writing, this="
                            << this << "\n";
       if (_shouldKeepAlive && this->_config._idleTimeout.count() > 0) {
         FUERTE_LOG_HTTPTRACE << "setting idle keep alive timer, this=" << this
                              << "\n";
-        setTimeout(this->_config._idleTimeout);
+        setTimeout(this->_config._idleTimeout, TimeoutType::IDLE);
       } else {
         this->shutdownConnection(Error::CloseRequested);
       }
       return;
     }
-    _active.store(true);
+    if (_active.exchange(true)) {
+      return;   // somebody else did this for us
+    };
+    _queue.pop(ptr);
   }
   this->_numQueued.fetch_sub(1, std::memory_order_relaxed);
 
   std::unique_ptr<http::RequestItem> item(ptr);
```

Comment:
there should be an assertion here that `ptr` is not a nullptr!

Reply:
Added.